### ISSUE TITLE
Request - HW Keyboard shortcut to change tabs

### ISSUE THREAD

**Issue by Shaolo** [omz/Pythonista-Issues#342]
It would be ideal if the keyboard shortcuts universally used to change tabs could be added for hardware keyboards.

Command-Shift-[ for previous tab
Command-Shift-] for next tab

**Comment by zrzka** [comment]
On iOS, Safari, ... these shortcuts are Ctrl-Tab and Ctrl-Shift-Tab.

**Comment by Shaolo** [comment]
Have you tried command shift bracket?

**Comment by omz** [comment]
There is currently no shortcut for switching tabs in Pythonista. When I add one, it'll probably be consistent with Safari. Cmd+Shift+Bracket has the additional disadvantage that it doesn't work on a lot of international keyboards (e.g. on a German keyboard, you need alt to type brackets, there's no dedicated key for them).

**Comment by Shaolo** [comment]
Perhaps both?

Both work in Safari.

Ctrl-Tab is not left hand friendly.

**Comment by zrzka** [comment]
Cmd-Shift-[ works well on Czech keyboard as well, but when I hold down Cmd, Safari shows me that I should use Ctrl-Tab and Ctrl-Shift-Tab.

**Comment by omz** [comment]
Didn't know that both work in Safari, makes sense.

**Comment by Shaolo** [comment]
I don't disagree.

@zrzka Really appreciate you writing and releasing Black Mamba btw. Made a huge difference in my ability to navigate the editor efficiently.

**Comment by omz** [comment]
@Shaolo Didn't think about left-hand-friendliness, good point!

**Comment by omz** [comment]
I think I should do it exactly like Safari. This way, users won't have to look it up if they already know the Safari shortcuts.

**Comment by dgelessus** [comment]
(I think I already suggested this somewhere else...) Firefox supports switching to tabs 1 through 9 using Cmd+1 through Cmd+9. I personally find that quite useful for navigating to a specific tab, it would be nice if Pythonista supported that too (in addition to normal next/previous tab switching).

**Comment by Shaolo** [comment]
@omz It's OK. No one ever thinks of us. 😭 

If you wanted to use the ctrl-tab for the keyboard popup reference, but capture both versions you would get no complaints from me. Safari supports both. I've never used ctrl-tab and I use the 🍎  pro keyboard.

**Comment by zrzka** [comment]
@Shaolo you're welcome. BTW this is another thing on my todo list. Next/prev tab & specific tab with Cmd-1 ... Cmd-9.

**Comment by omz** [comment]
@Shaolo Yeah, I think it makes sense to show only the tab variants in the popup reference. The API doesn’t support assigning multiple shortcuts to one command, so they’d internally be two separate shortcuts with the same action (and would therefore show up as duplicates in the popup reference).

**Comment by Shaolo** [comment]
@omz Check this out: https://developer.apple.com/documentation/uikit/uikeycommand/1621094-discoverabilitytitle

If the title is left blank it doesn't show up in the HUD.

🤞

**Comment by zrzka** [comment]
Yup, @Shaolo is right. Got some hidden shortcuts in Black Mamba as well :)

**Comment by omz** [comment]
@zrzka Since you worked on custom shortcuts as well, do you have any idea why I might not get Ctrl+Tab to work with `UIKeyCommand`? I don't have trouble adding other shortcuts, but I can't get this particular combination to work for some reason (Ctrl+Shift+Tab also doesn't work).

**Comment by zrzka** [comment]
@omz not sure. Apple reserves specific shortcuts? Will try later today. Workaround handleUIKeyEvent?

**Comment by omz** [comment]
@zrzka Forgive my ignorance, but what is `handleUIKeyEvent`?

**Comment by zrzka** [comment]
@omz it's actually `handleKeyUIEvent:` (`UIApplication`). I do use it for arrows, enter, shift enter, ... in my picker. See [key_events.py](https://github.com/zrzka/blackmamba/blob/master/blackmamba/key_events.py).

**Comment by zrzka** [comment]
And usage in [picker.py](https://github.com/zrzka/blackmamba/blob/master/blackmamba/picker.py#L151-L178).

**Comment by omz** [comment]
@zrzka Looks like this is private API, so I can't use it in Pythonista. 😞

**Comment by zrzka** [comment]
@omz aaargh, did forget it. Shame I can use anything I want, but you can't.

**Comment by zrzka** [comment]
Also it seems that there's no Tab in [Input Strings for Special Keys](https://developer.apple.com/documentation/uikit/uikeycommand/input_strings_for_special_keys?language=objc).

**Comment by omz** [comment]
@zrzka I tried `"\t"`, and it *seemed* to work, i.e. the `⇥` symbol showed up in the help popup (when holding Cmd key).

**Comment by zrzka** [comment]
@omz just tested it with Black Mamba and it works ...

```python
#!python3

import blackmamba as bm
bm.register_default_key_commands()

from blackmamba.key_commands import register_key_command
from blackmamba.uikit import *

def next_tab():
    print('Next Tab')


def previous_tab():
    print('Previous Tab')    
            
            
register_key_command('\t', UIKeyModifierControl, next_tab, 'Next tab')
register_key_command('\t', UIKeyModifierControl | UIKeyModifierShift, previous_tab, 'Next tab')
```

**Comment by omz** [comment]
Okay, thanks! I have a slight suspicion that it might have to do with the fact that I don't have an actual Bluetooth keyboard here, so I'm testing with Type2Phone and the iOS Simulator. Could be that not all hardware keyboard features are supported there.

**Comment by omz** [comment]
I just noticed that Ctrl+Tab doesn't work in Safari either, when using the iOS Simulator, so that's probably it. Sorry for bothering you with this.

**Comment by zrzka** [comment]
Yup, even the symbol is here as well :)

![5b34d362-a5eb-415b-a91c-d4b393878080](https://user-images.githubusercontent.com/1084172/29528807-9ba45a80-869d-11e7-9e16-3701a19272c6.jpeg)

These external keyboards are weird sometimes. Glad it helped.

**Comment by zrzka** [comment]
@omz I assume you're working on this feature, I can close https://github.com/zrzka/blackmamba/issues/4 and use my time for another Black Mamba feature?

**Comment by omz** [comment]
@zrzka Yep.

**Comment by zrzka** [comment]
@Shaolo in the meanwhile you can use [blackmamba-0.0.14](https://github.com/zrzka/blackmamba/blob/master/CHANGELOG.rst#0014-2017-08-25) which does contain all these shortcuts. Later on, when Pythonista will support them directly, I'll remove them from Black Mamba. Or better, will add condition to instantiate them only for older Pythonista versions.

**Comment by zrzka** [comment]
@omz there're shortcuts like Ctrl-Tab, Ctrl-Shift-Tab in the latest beta, but they do nothing. Black Mamba disabled.

**Comment by zrzka** [comment]
Same applies to Cmd-W, does nothing.

**Comment by zrzka** [comment]
Current status:

* Ctrl-Tab works in files
* Ctrl-Shift-Tab kinda works in files (see #444)
* Cmd-W works

What doesn't work and should be supported IMHO:

* Ctrl-[Shift]-Tab to switch between console, documentation, browser, ... tabs
* Cmd-W to close documentation, browser, ...

Safari compatibility:

* Safari does use Cmd-Shift-] for next tab, Cmd-Shift-[ for previous tab, these shortcuts are not there

To be honest, I do use Cmd-Shift-] more than Ctrl-Tab, because I'm not forced to move my fingers so far.

**Comment by omz** [comment]
I though Safari used `Cmd-[` and `Cmd-]` (without shift)? These shortcuts should be there (though not shown in the overview when holding the Cmd key).

**Comment by omz** [comment]
Ah, never mind. I just saw #446, so it seems Safari uses those for back/forward.

**Comment by zrzka** [comment]
Shortcuts work in the editor, but they don't in the accessories (console, help). Here's what happens:

* Go to editor
* Open two files (you have two tabs)
* Open console
* Hit `Cmd W`
* One of the files is closed in the editor
* Same applies to other shortcuts as well (tab switching).

I would expect that when I'm looking at accessories (help, console, panels, ...), that these shortcuts will work there as well.
